src/cargo/util/auth/mod.rs

```diff
@@ -440,285 +448,95 @@ pub fn auth_token(
 fn auth_token_optional(
     config: &Config,
     sid: &SourceId,
-    mutation: Option<&'_ Mutation<'_>>,
+    operation: Operation<'_>,
+    www_authenticate: Option<Vec<String>>,
 ) -> CargoResult<Option<Secret<String>>> {
+    log::debug!("token requested for {}", sid.display_registry_name());
     let mut cache = config.credential_cache();
     let url = sid.canonical_url();
-
-    if let Some(cache_token_value) = cache.get(url) {
-        // Tokens for endpoints that do not involve a mutation can always be reused.
-        // If the value is put in the cache by the command line, then we reuse it without looking at the configuration.
-        if cache_token_value.from_commandline
-            || cache_token_value.independent_of_endpoint
-            || mutation.is_none()
+    if let Some(cached_token) = cache.get(url) {
+        if cached_token
+            .expiration
+            .map(|exp| OffsetDateTime::now_utc() < exp)
```

Comment:
Does this need a grace window to handle the case where it expires in like 1 second? 
(Probably not too necessary, just curious if that needs to be considered.)